CIP-0165/README.md

```diff
@@ -0,0 +1,417 @@
+- `namespace` : `bstr` — namespace of the values stored in the CHUNK
+- `entries` — CBOR encoded list of unlimited length
+- `footer {entries_count: u64, chunk_hash: u64}` — hash value of the chunk of data, is used to keep integrity of the file.
+
+**Policy:**
+
+- chunk size \~8–16MiB; footer required;
+- data is stored in deterministically defined global order; In the lexical order of the keys;
+- all keys in the record must be unique;
+- all key-values in the record must refer to the same namespace;
+- readers should verify footer before relying on the data;
+- `chunk_hash = H(concat [ digest(e) | e in entries ])`;
+- all keys in `CHUNK` `n` must be lexicographically lower than all keys in `CHUNK` `n+1`.
+
+The format proposes support of data compression. For future-compatibility the format is described by the `chunk_format` field, and following variants are introduced:
+
+| Code | Name  | Description                                   |
+| ---- | ----- | --------------------------------------------- |
+| 0x00 | RAW   | Raw CBOR Entries                              |
+| 0x01 | ZSTD  | All entries are compressed with seekable zstd |
+| 0x02 | ZSTDE | Compress each value independently             |
+
+When calculating and verifying hashes, it's build over the uncompressed data.
+
+#### MANIFEST Record
+
+**Purpose:** index of chunks and information for file integrity check.
+
+**Structure:**
+
+- `total_entries`: `u64` — number of data entries in the file (integrity purpose only)
+- `total_chunks`: `u64` — number of chunks in the file (integrity purpose only)
+- `root_hash`: **Merkle root** of all `entry_e` in the chosen order, see verification for details
+- `namespace_hashes`: CBOR table of Merkle roots for each namespace, mapping namespace name to the hash in a multihash format
+- `prev_manifest_offset`: `u64` — offset of the previous manifest (used with delta files), zero if there is no previous manifest entry
+- `summary`: `{ created_at, tool, comment? }`
+
+**Policy:** used to verify all the chunks.
+
+#### DELTA Record
+
+```text
+TODO: exact contents of the DELTA record will be defined later, currently we describe
+a high-level proposal.
+```
+
+**Purpose:** Delta records are used to build iterative updates, when base format is created and we want to store additional transactions in a fast way. Delta records are designed to be compatible with UTxO-HD, LSM-Tree or other storage types where it's possible to stream list of updates.
+
+Updating of the file in-place is unsafe so instead we store list of updated.
+
+All updates are written in the following way:
+
+- if an entry has a natural key and we want to update data we can store a new value;
+- if we want to remove a value with natural key we should write a special tombstone entry, see namespaces;
+- if a value has no natural key we must first delete old value and then insert a new one.
+
+**Structure:**
+
+- `transaction:` `u64` — transaction number
```

Comment:
Transactions do not have numbers, nor are they stored in the ledger state

Reply:
Hm it was intended to use logical time, so it should be block number or slot number.
At this point I think that slot number may be better, but I'm not sure yet.